New collections picker

poltak opened the pull request:
Addresses part of #1006, implementing the new collections picker based on the tags picker work.

## TODOs:
- [x] BG custom-lists storage
  - [x] add missing types + set up RPCs with rest of newer RPC implementation
  - [x] add/update RPC methods for use with new UI
    - [x] list suggestions
    - [x] "recently-interacted-with" cache*
    - [x] get existing lists for page
    - [x] add all tabs to list*
    - [x] add/remove list for page
      - [x] ensure can work with just list names instead of IDs
  - [ ] cleanup existing storage methods 
    - [x] figure out what to do with twitter integration code (last place that uses `AddListDropdownContainer`) - **leave it alone**
      - [x] ~is it easy to replace use with new collections picker? if so, just replace it so `AddListDropdownContainer` becomes dead code~
    - [x] ~remove RPCs only used in `AddListDropdownContainer`~ - **flagged with `__` prefix for later removal once we completely drop this**
    - [x] assess state of remaining RPCs + storage methods
    - [x] ensure unit tests suck less (should incrementally improve as I update methods)
- [x] collection picker UI
  - [x] copy over tag picker source
  - [x] rename things appropriately
  - [x] start thinking about what could be extracted out and shared with tags picker + filter
    - [x] extract out presentational stuff
    - [x] extract out stateful logic
    - [x] ensure `DisplayEntry` type used instead of `DisplayTag`/`DisplayList`
  - [x] factor in differences
    - [x] figure out what the differences are
  - [x] integrate
    - [x] popup
    - [x] ribbon
    - [x] overview page results
    - [x] sidebar page results
- [x] address found bugs
  - [x] both pickers
    - [x] lots of duplicate entries appear in pickers
    - [x] searching shows results that aren't attached to the page or match the query
    - [x] mouse-over **some** entries no longer does anything after typing in some search query
  - [x] list picker exclusive
    - [x] lists added to a page appear twice
- [ ] extra
  - [x] New loading state views
    - [ ] fix up
  - [x] Show a "no tags yet" screen when no tags are there. but also only after the loading is complete.
  - [x] new icon for search results collections picker show button

\* _possible candidates for extracting logic out and sharing with tags picker and/or filter (more just making notes of things to think more about later)_

poltak commented:
>  figure out what the [UI] differences are

@blackforestboi it would be good to clarify these reqs now. What are the visual differences between the tags and collections picker? I couldn't find any existing mockup.

For ref, the only differences between the old tags and coll picker that I could see was the absence of the blue "pill" styles on entries in the collections picker.

blackforestboi commented:
> @blackforestboi it would be good to clarify these reqs now. What are the visual differences between the tags and collections picker? I couldn't find any existing mockup.

They are here: https://www.figma.com/file/Ec1NJKz82yAJ2MEXipap0K/Tags?node-id=380%3A0

blackforestboi commented:
Works wonderfully. So looking forward having those 2 updates finally in! 

There is one bug though: When adding a new collection (existing installation) it adds them twice to the list of collections in the picker (but not in the actual collections list in the dashboard). Re-adding an existing collection adds it AGAIN to the list. so many duplicates appear. 
https://recordit.co/ut5bOMeJBV

And one feature we still need: A dedicated icon to add a page to a collection in the dashboard and in-page sidebar results.
<img width="852" alt="Screenshot 2020-04-27 at 16 44 17" src="https://user-images.githubusercontent.com/7870039/80385480-62195f00-88a6-11ea-872b-f699c2dc4146.png">

blackforestboi commented:
Ok, now that I am styling things I am running into tons of weird bugs. 
These are universal between tag picker and collections picker though. 

1. Probably related to the bugs in the previous post: Searching for a tag still shows some other results even though they should not be there. <img width="313" alt="Screenshot 2020-04-27 at 21 25 08" src="https://user-images.githubusercontent.com/7870039/80412308-f814b000-88cd-11ea-9dd9-89262091a6f7.png">
2. This one seems to be related to 1) When typing into the search field, my mouseover the results doesn't trigger anymore on most of the results - those that should not be there in particular. Also I can't use the keyboard arrows anymore to navigate the list. Also the mouseover causes some weird glitches.http://recordit.co/51Jq5GZ0B5
3. This one is only for collections picker, and probably related to 1 too: When searching for a collection that is already existing. It does not show it, just some weird results. The highest ranked search should always be the first result
, and also should be in focus so that a 'enter' key confirms them. Now it just freezes in that view and pressing 'enter' doesnt do anything. <img width="320" alt="Screenshot 2020-04-27 at 21 24 24" src="https://user-images.githubusercontent.com/7870039/80412477-38742e00-88ce-11ea-9692-e086f1740e00.png">

Other considerations: 
- Show the loading circles for the entire box including the search box for the state where existing tags are loaded. Especially with many tags/collections in the system things can take a bit for first load - its just better ux to know things are loading instead of staring at an empty list expecting things to be there. 
- Show a "no tags yet" screen when no tags are there. but also only after the loading is complete.

blackforestboi commented:
I also changed all the tag mentions into "list". 
where I wasn't sure about: 
- What to do with the theme props like `props.theme.tag.text` 

Now there is one test failing - but seems to be minor. Build works. 

@kellective @cdharris re the search icon: I removed it all together and it looks really good this way. 
<img width="325" alt="Screenshot 2020-04-27 at 23 10 25" src="https://user-images.githubusercontent.com/7870039/80421162-521c7200-88dc-11ea-9df1-1ca3dcc76538.png">
<img width="330" alt="Screenshot 2020-04-27 at 23 10 31" src="https://user-images.githubusercontent.com/7870039/80421161-50eb4500-88dc-11ea-951e-49d809a3d89b.png">

poltak commented:
> - [ ] start thinking about what could be extracted out and shared with tags picker + filter

I just got to this task (relates to the 3rd option expressed in #1006) though I soon realized that: apart from @blackforestboi's changes in 348f0bbe18e95d3d3798fd2dc9dc5ebb96d2f971, there hasn't been any need to touch the new UI code. Everything works well as is.
i.e., everything in `src/custom-lists/ui/CollectionsPicker` is just a copy and paste of `src/tags/ui/TagsPicker` apart from some styles (components) and namings (components + logic).

This makes a lot of sense really. I haven't looked into these noted bugs yet, though I'm guessing a lot of them are going to lead to duplicate fixes needing to be implemented in both codebases. This may make things more obvious in the way of figuring out what makes sense to share.

The possible logic and components I'm currently feeling may make sense to extract out and share so far (probably will update these lists as this becomse more clear) are:
components:
- [ ] the search input
- [ ] list of entries

logic (state + mutators):
- [ ] the search input
- [ ] the entries

I'm really not too sure on the extent to go here, as almost everything could be shared. Hoping that becomes more clear as I jump into the "how" part of this task for these first few candidates

poltak commented:
Things became a lot more clear with what can be extracted out as I dug into the UI code. I have extracted out the majority of presentational components into `src/common-ui/GenericPicker/components/` as they were identical. Mostly done in e60fe9a and a bit in a687b8e. I still want to think about and attempt extracting out redundant stateful logic, which I'll do a bit later today.

@blackforestboi the bugs you reported have been taken care of and also added in the extra features. 
Some things to address:
1. The new loader for the search bar may want further styling. To easily get it to show, you can set this `loading` prop to `true`: https://github.com/WorldBrain/Memex/blob/e60fe9a2bc3d759bcec11e7e90818d2b3a2355a1/src/tags/ui/TagPicker/index.tsx#L109
2. The "show collections picker" button in search results needs differentiating from the tags button - for now I've just duplicated the icon
3. I feel it's too difficult to tell which collections in the picker are already added to the page and which aren't because there's no selected state styles differentiation. If you want to figure this out, you need to compare the collection names with those that show up above in the search bar, which is a bit slow especially compared with how it's done in the tags picker.

blackforestboi commented:
Sorry I have been a bit unprecise in the specs for the loader. 
The problem is that, especially if there is a lot of data to fetch or the system is under load, that there is a brief moment where potentially nothing is shown in the list. That can cause confusion because sometimes it can take a while. 
Showing the "no collections available" text then would be counterproductive. 

So by default on every tag/collections picker load it should show a loading bar instead of the results that only goes away if results are loaded in, or it is clear that there are no results (because new installation for example). The loader also shows up when doing a new query or so. 
It should look like this. The loader should also not be in the search bar - i know thats the current specs but its inconsistent with the way we have the loader elsewhere in the app. 

<img width="789" alt="Screenshot 2020-04-29 at 13 54 39" src="https://user-images.githubusercontent.com/7870039/80593294-3bcbfe80-8a21-11ea-8965-07bc0e61419a.png">

poltak commented:
@blackforestboi I have set up this new loading view as described in your last comment. I feel it's fine for the initial load but for loads on queries I have the issue that:
When the DB is fast to query, the UI flashes in and out all the time if you type slow - this is quite annoying. 

Maybe one approach could be to increase the debounce time (currently 150ms). 
Another way I tried was to keep the search bar visible for the query load state, but hidden for the inital load state. I still found it annoying with the flashing results list though. 

Maybe having everything still shown but unable to be interacted with during these 2 load states could be better?

Have a go to see what I mean with the current state, but I feel like you might need to put some more thought into this loading state

blackforestboi commented:
> When the DB is fast to query, the UI flashes in and out all the time if you type slow - this is quite annoying.

Yeah tested it now its super annoying. 

The loader in the screenshot should only appear on first load when it initialises. 
For all other loads it should look like this: 
<img width="789" alt="Screenshot 2020-04-29 at 13 54 39" src="https://user-images.githubusercontent.com/7870039/80689549-d124cc80-8acd-11ea-9ab7-51100cd63ef5.png">

is that what you describe here? 
> Another way I tried was to keep the search bar visible for the query load state, but hidden for the inital load state. I still found it annoying with the flashing results list though.

Maybe we actually don't need a loading state except for the first load - would that solve the problem?

poltak commented:
> is that what you describe here?

Yes, that's what I was meaning there and what I thought still felt annoying after trying it out.

> Maybe we actually don't need a loading state except for the first load - would that solve the problem?

That would solve this problem, and I think it's the easiest solution for now, but it may not solve (part of) the original problem you were intending to solve with the state in the first place. i.e., when it takes some time to search.

IMO for the search loading state: it would be nice to have the loading indicator somewhere **and** also not need to hide any elements of the UI.

blackforestboi commented:
That would solve this problem, and I think it's the easiest solution for now, but it may not solve (part of) the original problem you were intending to solve with the state in the first place. i.e., when it takes some time to search.

No my original problem was more of the time it takes to load the initial suggestions and the existing tags. That can sometimes take some time. 

Maybe then the best solution would be the one we had originally: 
1) Show the loader over everything on first load
2) show a loader in the search bar if it is loading suggestions

<img width="689" alt="Screenshot 2020-04-30 at 10 55 17" src="https://user-images.githubusercontent.com/7870039/80691892-244c4e80-8ad1-11ea-848e-6f066bf4e3be.png">
